### ISSUE TITLE
NamedTuples changed iteration behavior for vararg keywords

### ISSUE THREAD

**Issue by rdeits** [JuliaLang/julia#25177]
This seems to be a consequence of #24795 

In previous Julia versions, we could iterate through vararg keywords as key-value tuples:

```julia
julia> function foo(;kw...)
         for (k, v) in kw
           @show (k, v)
         end
       end
foo (generic function with 1 method)

julia> foo(x=1, y=2)
(k, v) = (:x, 1)
(k, v) = (:y, 2)
```

but after #24795, iterating through vararg keywords only gives the values:

```julia
julia> function foo(;kw...)
         for (k, v) in kw
           @show (k, v)
         end
       end
foo (generic function with 1 method)

julia> foo(x=1, y=2)
ERROR: BoundsError: attempt to access 1
  at index [2]
Stacktrace:
 [1] indexed_next at ./tuple.jl:60 [inlined]
 [2] #foo#2(::NamedTuple{(:x, :y),Tuple{Int64,Int64}}, ::Function) at ./REPL[3]:2
 [3] (::getfield(, Symbol("#kw##foo")))(::NamedTuple{(:x, :y),Tuple{Int64,Int64}}, ::typeof(foo)) at ./<missing>:0
 [4] top-level scope
```

If this is an intended change, then it should probably at least be mentioned in NEWS.md. I realize that the changelog mentions that "name-value pairs can be iterated using pairs(kw)" but I think it's important to clarify that this is a breaking change.

**Comment by rdeits** [comment]
It looks like the correct way to do this in a cross-version way is:

```julia
using Compat: pairs

function foo(;kw...)
  for (k, v) in pairs(kw)
    @show (k, v)
  end
end
```

**Comment by vtjnash** [comment]
Perhaps we should call the inner function with the object `kw = pairs(kw')`, so that there's no change to the API, just faster kwargs?

**Comment by rdeits** [comment]
That would definitely make adapting existing code easier.

**Comment by andyferris** [comment]
I think this was known, but it's annoying as a breaking change.

Wrapping in `pairs` would destroy the ability to getfield the kwargs, no?

**Comment by vtjnash** [comment]
no, it'll still maintain it's new abilities from now being a NamedTuple

**Comment by StefanKarpinski** [comment]
That does not seem to be true at all, @vtjnash:
```jl
julia> f(; kws...) = pairs(kws)
f (generic function with 1 method)

julia> kws = f(a = 1, b = 2)
Base.Generator{Base.Iterators.Zip2{Tuple{Symbol,Symbol},Tuple{Int64,Int64}},getfield(Base, Symbol("##7#8")){Pair}}(getfield(Base, Symbol("##7#8")){Pair}(), Base.Iterators.Zip2{Tuple{Symbol,Symbol},Tuple{Int64,Int64}}((:a, :b), (1, 2)))

julia> kws.a
ERROR: type Generator has no field a
```

**Comment by vtjnash** [comment]
You don’t seem to be on master

**Comment by StefanKarpinski** [comment]
I'm a day-old version of master so unless this is more recent than that.

**Comment by StefanKarpinski** [comment]
Fresh master has the same error, although it does print the key-value map in the pairs object:
```jl
julia> f(; kws...) = pairs(kws)
f (generic function with 1 method)

julia> kws = f(a = 1, b = 2)
Base.Iterators.IndexValue{Symbol,Int64,Tuple{Symbol,Symbol},NamedTuple{(:a, :b),Tuple{Int64,Int64}}} with 2 entries:
  :a => 1
  :b => 2

julia> kws.a
ERROR: type IndexValue has no field a
Stacktrace:
 [1] getproperty(::Base.Iterators.IndexValue{Symbol,Int64,Tuple{Symbol,Symbol},NamedTuple{(:a, :b),Tuple{Int64,Int64}}}, ::Symbol) at ./sysimg.jl:8
 [2] top-level scope
```

**Comment by vtjnash** [comment]
`kws[:a]` – `getproperty` overloading only just merged yesterday, so we haven't yet had time to implement that API

**Comment by c42f** [comment]
This was fixed by #25290
